src/key.cpp

```diff
@@ -85,31 +79,39 @@ const char* modifiermask2name(unsigned int mask) {
     return nullptr;
 }
 
-int keybind(int argc, char** argv, Output output) {
-    if (argc <= 2) {
+int keybind(Input input, Output output) {
+    if (input.size() < 2) {
         return HERBST_NEED_MORE_ARGS;
     }
-    KeyBinding new_bind;
-    // get keycode
-    if (!string2key(argv[1], &(new_bind.modifiers), &(new_bind.keysym))) {
-        output << argv[0] << ": No such KeySym/modifier\n";
+
+    auto newBinding = make_unique<KeyBinding>();
+
+    // Extract modifiers/keysym
+    if (!string2key(input.front(), &(newBinding->modifiers), &(newBinding->keysym))) {
+        output << input.command() << ": No such KeySym/modifier\n";
         return HERBST_INVALID_ARGUMENT;
     }
-    KeyCode keycode = XKeysymToKeycode(g_display, new_bind.keysym);
+
+    // Validate keysym
+    KeyCode keycode = XKeysymToKeycode(g_display, newBinding->keysym);
     if (!keycode) {
-        output << argv[0] << ": No keycode for symbol "
-               << XKeysymToString(new_bind.keysym) << std::endl;
+        output << input.command() << ": No keycode for symbol "
+               << XKeysymToString(newBinding->keysym) << std::endl;
         return HERBST_INVALID_ARGUMENT;
     }
-    // remove existing binding with same keysym/modifiers
-    key_remove_bind_with_keysym(new_bind.modifiers, new_bind.keysym);
-    // create a copy of the command to execute on this key
-    new_bind.cmd_argc = argc - 2;
-    new_bind.cmd_argv = argv_duplicate(new_bind.cmd_argc, argv+2);
-    // add keybinding
-    g_key_binds.push_back(new_bind);
-    // grab for events on this keycode
-    grab_keybind(&g_key_binds.back());
+
+    input.shift();
+    // Store remaining input as the associated command
+    newBinding->cmd = {input.begin(), input.end()};
+
+    // Remove existing binding with same keysym/modifiers
+    key_remove_bind_with_keysym(newBinding->modifiers, newBinding->keysym);
+
+    // Grab for events on this keycode
+    grab_keybind(newBinding.get());
+
+    // Add keybinding to list
+    g_key_binds.push_back(std::move(newBinding));
```

Comment:
Please note that I've swapped the order of `grab_keybind` and `push_back` (because it made things easier). I'm assuming that it doesn't make a difference.

Reply:
Is fine I think.